src/core/license.rs

```diff
@@ -63,82 +96,914 @@ pub enum StateTransition {
+        _ => {
+            return Err(Error::RdpError(RdpError::new(
+                RdpErrorKind::InvalidRespond,
+                "unexpected license message",
+            )))
+        }
+    }
+
+    let (_channel, payload) = mcs.read()?;
+    match LicenseMessage::new(payload)? {
+        LicenseMessage::NewLicense(new_license) => {
+            // At the moment we're not storing license client-side,
+            // so we can just validate data and return
+            let _license = License::new(&session_encryption_data, &new_license)?;
+        }
+        LicenseMessage::ErrorAlert(error_alert) => return error_alert.is_valid(),
+        _ => {
+            return Err(Error::RdpError(RdpError::new(
+                RdpErrorKind::InvalidRespond,
+                "unexpected license message",
+            )))
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const CLIENT_RANDOM_BUFFER: [u8; 32] = [
+        29, 91, 101, 210, 58, 6, 253, 117, 36, 209, 128, 84, 229, 91, 28, 208, 0, 65, 109, 164,
+        235, 10, 22, 69, 183, 184, 158, 209, 128, 157, 122, 187,
+    ];
+    const SERVER_RANDOM_BUFFER: [u8; 32] = [
+        135, 43, 167, 218, 195, 159, 70, 134, 193, 59, 40, 249, 168, 129, 51, 230, 69, 18, 45, 19,
+        183, 23, 238, 173, 95, 50, 156, 45, 254, 174, 194, 255,
+    ];
+    const ENCRYPTED_PREMASTER_SECRET: [u8; 72] = [
+        62, 187, 238, 159, 188, 213, 152, 124, 16, 119, 43, 178, 109, 127, 28, 8, 220, 233, 172,
+        132, 171, 132, 190, 12, 152, 243, 248, 102, 237, 102, 219, 221, 138, 100, 200, 255, 1, 223,
+        23, 145, 121, 223, 164, 9, 221, 186, 171, 182, 80, 28, 207, 0, 78, 51, 90, 129, 249, 53,
+        187, 221, 207, 31, 213, 62, 0, 0, 0, 0, 0, 0, 0, 0,
+    ];
+    const PREMASTER_SECRET_BUFFER: [u8; 48] = [
+        10, 63, 107, 103, 158, 17, 153, 19, 221, 157, 28, 215, 121, 50, 89, 79, 194, 171, 104, 34,
+        180, 124, 248, 187, 66, 180, 163, 60, 208, 90, 236, 216, 4, 98, 137, 236, 151, 108, 126,
+        215, 164, 2, 191, 110, 236, 93, 51, 40,
+    ];
+    const SALTED_HASH_BUFFER: [u8; 16] = [
+        121, 64, 28, 102, 24, 136, 13, 198, 158, 252, 57, 55, 222, 122, 4, 93,
+    ];
+    const MASTER_SECRET_BUFFER: [u8; 48] = [
+        121, 64, 28, 102, 24, 136, 13, 198, 158, 252, 57, 55, 222, 122, 4, 93, 193, 218, 164, 59,
+        153, 133, 183, 145, 40, 134, 109, 239, 233, 122, 214, 75, 248, 66, 147, 166, 135, 133, 144,
+        178, 145, 156, 51, 212, 87, 202, 238, 190,
+    ];
+    const SESSION_KEY_BLOB: [u8; 48] = [
+        116, 192, 213, 13, 126, 75, 71, 132, 142, 4, 227, 137, 9, 236, 124, 46, 180, 3, 198, 101,
+        137, 71, 69, 8, 249, 165, 203, 120, 82, 245, 219, 102, 102, 5, 158, 87, 22, 221, 137, 195,
+        241, 120, 125, 127, 224, 14, 174, 47,
+    ];
+    const LICENSE_KEY_BUFFER: [u8; 16] = [
+        44, 229, 51, 182, 241, 98, 94, 249, 201, 240, 209, 190, 72, 113, 246, 41,
+    ];
+    const X509_CERTIFICATE: [u8; 1117] = [
```

Comment:
Maybe we don't care since this has a uh, healthy expiration date (2038) and probably won't need to be rotated any time soon, but is this something we might want to include from a file via `include_bytes!()`? If nothing else it's much easier to inspect and debug with `openssl` as a .der.

Reply:
Yeah we don't care that much about the certificate information in this case as we only care about public key in this tests. Anyway, I'll check that `include_bytes!` macro.